framework/aster-frame/src/softirq/tasklet.rs

```diff
@@ -0,0 +1,169 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::sync::Arc;
+use core::{
+    cell::RefCell,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+use intrusive_collections::{intrusive_adapter, LinkedList, LinkedListAtomicLink};
+
+use super::raise_softirq;
+use crate::{
+    arch::irq::{disable_local, enable_local},
+    cpu_local,
+    sync::SpinLock,
+    CpuLocal,
+};
+
+enum TaskLetState {
+    Free,      // The tasklet is free and can be scheduled to any tasklet list.
+    Scheduled, // The tasklet has been scheduled in a tasklet list.
+    Running,   // The tasklet is running.
+}
+
+/// TaskLet mechanism.
+///
+/// # Overview
+///
+/// `TaskLet` provides an abstraction for deferring execution of a small task that
+/// does not need to sleep and can be executed in the interrupt context. It is primarily
+/// used for tasks that cannot be executed directly in the interrupt handler due to
+/// their potentially longer execution time or non-atomic nature.
+///
+/// # Examples
+///
+/// Users can create a `TaskLet` and schedule it at any place.
+/// ```rust
+/// #use aster_frame::softirq::TaskLet;
+///
+/// #fn my_func() {}
+///
+/// let tasklet = TaskLet::new(Arc::new(my_func));
+/// // This tasklet will be executed in softirq context soon.
+/// schedule_tasklet(&tasklet);
+///
+/// ```
+/// Note that the action function should not involve sleep instructions.
+pub struct TaskLet {
+    /// Current state of this tasklet.
+    state: SpinLock<TaskLetState>,
+    link: LinkedListAtomicLink,
+    /// The function that will be called when execute this tasklet.
+    action: Arc<dyn Fn() + Send + Sync + 'static>,
+    /// If this tasklet is disabled.
+    is_disabled: AtomicBool,
+}
+
+impl TaskLet {
+    /// Create a new `TaskLet` instance with its action.
+    pub fn new<F>(action: Arc<F>) -> Arc<Self>
```

Comment:
If no reason to have multiple owners, then takes an `action: F` argument and turns it into a `Box` internally.